File: bundler-compiler/src/main/java/com/hendraanggrian/bundler/compiler/BundlerProcessor.java
package com.hendraanggrian.bundler.compiler;

import com.google.auto.common.MoreElements;
import com.google.auto.service.AutoService;
import com.google.common.collect.ImmutableSet;
import com.google.common.collect.LinkedHashMultimap;
import com.google.common.collect.Multimap;
import com.google.common.collect.Sets;
import com.hendraanggrian.bundler.BindExtra;
import com.hendraanggrian.bundler.BindState;
import com.squareup.javapoet.JavaFile;

import java.io.IOException;
import java.lang.annotation.Annotation;
import java.util.Set;

import javax.annotation.processing.AbstractProcessor;
import javax.annotation.processing.Filer;
import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.Processor;
import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.Element;
import javax.lang.model.element.TypeElement;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
@AutoService(Processor.class)
public final class BundlerProcessor extends AbstractProcessor {

    private static final Set<Class<? extends Annotation>> SUPPORTED_ANNOTATIONS = ImmutableSet.<Class<? extends Annotation>>of(
            BindExtra.class,
            BindState.class
    );

    private Types typeUtils;
    private Elements elementUtils;
    private Filer filer;

    @Override
    public SourceVersion getSupportedSourceVersion() {
        return SourceVersion.latestSupported();
    }

    @Override
    public Set<String> getSupportedAnnotationTypes() {
        Set<String> names = Sets.newLinkedHashSet();
        for (Class<? extends Annotation> cls : SUPPORTED_ANNOTATIONS) {
            names.add(cls.getCanonicalName());
        }
        return names;
    }

    @Override
    public synchronized void init(ProcessingEnvironment processingEnv) {
        super.init(processingEnv);
        typeUtils = processingEnv.getTypeUtils();
        elementUtils = processingEnv.getElementUtils();
        filer = processingEnv.getFiler();
    }

    @Override
    @SuppressWarnings("ResultOfMethodCallIgnored")
    public boolean process(Set<? extends TypeElement> set, RoundEnvironment roundEnv) {
        // build utility class if parceler is available and if it has not yet already been created
        if (elementUtils.getTypeElement(Spec.CLASS_PARCELS.toString()) != null &&
                elementUtils.getTypeElement(Spec.CLASS_BUNDLER_UTILS.toString()) == null) {
            JavaFile file = new UtilsSpec().toJavaFile();
            try {
                file.writeTo(filer);
            } catch (IOException ignored) {
            }
        }
        // preparing elements
        for (Class<? extends Annotation> annotation : SUPPORTED_ANNOTATIONS) {
            String suffix = annotation == BindExtra.class
                    ? BindExtra.SUFFIX
                    : BindState.SUFFIX;
            Multimap<TypeElement, Element> map = LinkedHashMultimap.create();
            Set<String> generatedClassNames = Sets.newHashSet();
            for (Element fieldElement : roundEnv.getElementsAnnotatedWith(annotation)) {
                TypeElement typeElement = MoreElements.asType(fieldElement.getEnclosingElement());
                map.put(typeElement, fieldElement);
                generatedClassNames.add(Spec.guessGeneratedName(typeElement, suffix));
            }
            // write classes and keep results
            for (TypeElement key : map.keySet()) {
                JavaFile file = new BindingSpec(key, suffix)
                        .superclass(generatedClassNames)
                        .statement(map.get(key), typeUtils)
                        .toJavaFile();
                try {
                    file.writeTo(filer);
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            }
        }
        return false;
    }
}

File: bundler-compiler/src/main/kotlin/com/hendraanggrian/bundler/compiler/BundlerProcessor.kt
package com.hendraanggrian.bundler.compiler

import com.google.auto.common.MoreElements
import com.google.common.collect.LinkedHashMultimap
import com.google.common.collect.Sets
import com.hendraanggrian.bundler.Extra
import com.hendraanggrian.bundler.State
import java.io.IOException
import javax.annotation.processing.AbstractProcessor
import javax.annotation.processing.Filer
import javax.annotation.processing.ProcessingEnvironment
import javax.annotation.processing.RoundEnvironment
import javax.lang.model.SourceVersion
import javax.lang.model.element.Element
import javax.lang.model.element.TypeElement
import javax.lang.model.util.Elements
import javax.lang.model.util.Types

class BundlerProcessor : AbstractProcessor() {

    companion object {
        private val SUPPORTED_ANNOTATIONS = setOf(Extra::class.java, State::class.java)
    }

    private lateinit var mTypeUtils: Types
    private lateinit var mElementUtils: Elements
    private lateinit var mFiler: Filer

    override fun getSupportedSourceVersion(): SourceVersion = SourceVersion.latestSupported()
    override fun getSupportedAnnotationTypes(): Set<String> = SUPPORTED_ANNOTATIONS.map { it.canonicalName }.toSet()

    @Synchronized override fun init(processingEnv: ProcessingEnvironment) {
        super.init(processingEnv)
        mTypeUtils = processingEnv.typeUtils
        mElementUtils = processingEnv.elementUtils
        mFiler = processingEnv.filer
    }

    override fun process(set: Set<TypeElement>, roundEnv: RoundEnvironment): Boolean {
        // build utility class if parceler is available and if it has not yet already been created
        if (mElementUtils.getTypeElement(TYPE_PARCELS.toString()) != null && mElementUtils.getTypeElement(TYPE_BUNDLER_UTILS.toString()) == null) {
            val file = UtilsSpec().toJavaFile()
            try {
                file.writeTo(mFiler)
            } catch (ignored: IOException) {
            }
        }
        // preparing elements
        for (annotation in SUPPORTED_ANNOTATIONS) {
            val map = LinkedHashMultimap.create<TypeElement, Element>()
            val generatedClassNames = Sets.newHashSet<String>()
            for (fieldElement in roundEnv.getElementsAnnotatedWith(annotation)) {
                val typeElement = MoreElements.asType(fieldElement.enclosingElement)
                map.put(typeElement, fieldElement)
                generatedClassNames.add(typeElement.getMeasuredName(if (annotation == Extra::class.java) Extra.SUFFIX else State.SUFFIX))
            }
            // write classes and keep results
            map.keySet()
                    .map {
                        (if (annotation == Extra::class.java) ExtraBindingSpec(it)
                        else StateBindingSpec(it))
                                .superclass(generatedClassNames)
                                .statement(map.get(it), mTypeUtils)
                                .toJavaFile()
                    }
                    .forEach {
                        try {
                            it.writeTo(mFiler)
                        } catch (e: IOException) {
                            throw RuntimeException(e)
                        }
                    }
        }
        return false
    }
}